File: constructi-backend/src/main/java/com/constructi/service/TaskService.java
package com.constructi.service;


import com.constructi.DTO.TaskRequestDTO;
import com.constructi.DTO.TaskResponseDTO;
import com.constructi.model.entity.Task;

import java.time.LocalDate;
import java.util.List;

public interface TaskService {

    TaskResponseDTO createTask(TaskRequestDTO taskRequestDTO);

    TaskResponseDTO updateTask(Long taskId, TaskRequestDTO taskRequestDTO);

    void deleteTask(Long taskId);

    List<TaskResponseDTO> getAllTasks();

    TaskResponseDTO getTaskById(Long taskId);

    List<TaskResponseDTO> getMyTasks();

    TaskResponseDTO assignTaskToWorker(Long taskId, Long workerId);

    List<TaskResponseDTO> getTasksAssignedToWorker();

//    TaskResponseDTO getTaskWithBudgetAndInvoices(Long taskId);

    TaskResponseDTO getTaskWithInvoices(Long taskId);

    void updateParentTaskStatusIfSubtasksCompleted(Long parentTaskId);


//    TaskResponseDTO prolongTask(Long taskId, LocalDate newEndDate);
}

File: constructi-frontend/src/app/modules/task/task.service.ts
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { HttpClient } from '@angular/common/http';
import {environment} from '../../../environments/environment';

@Injectable({
  providedIn: 'root'
})
export class TaskService {
  private apiUrl = environment.tasksApiUrl;

  constructor(private http: HttpClient) {}

  createTask(task: any): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/add`, task);
  }

  getAllTasks(): Observable<any[]> {
    return this.http.get<any[]>(`${this.apiUrl}`);
  }

  getTaskById(taskId: number): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/${taskId}`);
  }

  updateTask(taskId: number, task: any): Observable<any> {
    return this.http.put<any>(`${this.apiUrl}/update/${taskId}`, task);
  }

  deleteTask(taskId: number): Observable<string> {
    return this.http.delete(`${this.apiUrl}/delete/${taskId}`, { responseType: 'text' });
  }

  assignTaskToWorker(taskId: number, workerId: number): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/assign/${taskId}/${workerId}`, {});
  }

  getAssignedTasks(): Observable<any[]> {
    return this.http.get<any[]>(`${this.apiUrl}/mytasks`);
  }

  getTaskWithInvoices(taskId: number): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/invoice/${taskId}`);
  }

  // prolongTask(taskId: number, newEndDate: string): Observable<any> {
  //   return this.http.post<any>(`${this.apiUrl}/prolong/${taskId}`, null, {
  //     params: { newEndDate }
  //   });
  // }
}
